lib/include/openamp/virtqueue.h

```diff
@@ -48,9 +48,13 @@ extern "C" {
 /* Support to suppress interrupt until specific index is reached. */
 #define VIRTIO_RING_F_EVENT_IDX        (1 << 29)
 
+/* cache invalidation helpers */
+#define CACHE_FLUSH(x, s)		metal_cache_flush(x, s)
+#define CACHE_INVALIDATE(x, s)		metal_cache_invalidate(x, s)
+
 #ifdef VIRTIO_CACHED_VRINGS
-#define VRING_FLUSH(x, s)		metal_cache_flush(&x, s)
-#define VRING_INVALIDATE(x, s)		metal_cache_invalidate(&x, s)
+#define VRING_FLUSH(x, s)		CACHE_FLUSH(x, s)
+#define VRING_INVALIDATE(x, s)		CACHE_INVALIDATE(x, s)
 #else
 #define VRING_FLUSH(x, s)		do { } while (0)
 #define VRING_INVALIDATE(x, s)		do { } while (0)
```

Comment:
but why having the double macros here?

Why we cannot have something like:
```c
#ifdef USE_DCACHE
#define CACHE_FLUSH(x, s) ...
#define CACHE_INVALIDATE(x, s) ...
#else
...
#endif
```

and just use `CACHE_FLUSH` and `CACHE_INVALIDATE` when needed?

@arnopo do we really need to differentiate between cache operations on VRIGNS, BUFFERS, etc...?

Reply:
From [here](https://github.com/OpenAMP/open-amp/pull/500/files#r1261491593), I understand that we want to keep all 3 options and add a new one to enable all the other 3. That's why I kept all: [[1](https://github.com/OpenAMP/open-amp/pull/502/files#diff-066c419494ef83ff624e9dbbd5929144f4a5128d3335087aac84644ba4ae38f8R55)], [[2](https://github.com/OpenAMP/open-amp/pull/502/files#diff-03144eb025fe729b04676264f230f5ba70dcb5151263c7ce17076b154a53c841R33)]. [[3](https://github.com/OpenAMP/open-amp/pull/502/files#diff-de793c7a31ee3b88aae02e85c5b992259cf736e05ac4efd2b066eeec28c3bd8fR28)]

IMO we should keep also the 3 options separately, since the rings, buffers and resource table might be in different memory sections w/o cache.

But, @arnopo please clarify.

Reply:
done